luminous-lightyears/src/weighted_random.py

```diff
@@ -0,0 +1,51 @@
+"""Weight random list implementation."""
+
+import random
+from typing import TYPE_CHECKING, Generic, Protocol, TypeVar
+
+if TYPE_CHECKING:
+    from src.player import PlayerState
+
+
+class SupportedRandomValue(Protocol):
+    """Generic type for supported classes."""
+
+    @property
+    def weight(self) -> int: ...  # noqa: D102
+
+    def is_available(self, state: "PlayerState") -> bool: ...  # noqa: D102
+
+
+T = TypeVar("T", bound=SupportedRandomValue)
+
+
+class WeightedList(Generic[T]):
```

Comment:
I really like the use of type generics + protocol typing here. I am not too familiar with this; however it is a good use case and offers utility in my IDE that adds clarity to the codebase.